### PR TITLE
feat(ingest/vertica): performance improvement and bug fixes

### PR DESCRIPTION
**Pull request by vishalkSimplify**
## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

1.  Rearchitected the way queries were being executed to minimize the total number of queries during ingestion. Customers with large catalogs were seeing unacceptable performance. Changed to schema-level capture of information into memory and then iterating through memory to get individual object details. We did testing on our side and with new changes performance improvement was above 90% as compared to the already existing Vertica plugin.

2. We have removed Oauth metadata features from Vertica as it was exposing security-related pieces of information.

3. Added integration test for Vertica plugin which covers all our features per datahub request.

4. we added form-based UI ingestion for the Vertica plugin.

5. Upgraded vertica dialect from 0.0.1 to 0.0.8 which now supports the latest sqlalchemy features.

6. Bug fixes and other small improvements.

### CODE REVIEW

File: datahub-web-react/src/app/ingest/source/builder/RecipeForm/constants.ts
Comment (vishalkSimplify) [inline datahub-web-react/src/app/ingest/source/builder/RecipeForm/constants.ts:1]: added support for ingest from form in ui

File: metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py [hunk 1]
```diff
@@ -290,29 +278,43 @@ def loop_views(
...
-                            yield MetadataChangeProposalWrapper(
+                            upstream_lineage = lineage_info
+
+                            lineage_mcpw = MetadataChangeProposalWrapper(
                                 entityType="dataset",
                                 changeType=ChangeTypeClass.UPSERT,
                                 entityUrn=dataset_snapshot.urn,
                                 aspectName="upstreamLineage",
-                                aspect=lineage_info,
-                            ).as_workunit()
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:315]: revert this change

we no longer need to call report_workunity and `.as_workunit()` is cleaner

File: metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py [hunk 2]
```diff
@@ -349,25 +351,104 @@ def _process_view(
...
         )
-        view_owner = self._get_owner_information(view, "view")
+
+        view_owner = inspector.get_view_owner(view, schema)
         yield from add_owner_to_entity_wu(
             entity_type="dataset",
             entity_urn=dataset_urn,
             owner_urn=f"urn:li:corpuser:{view_owner}",
         )
-        yield from super()._process_view(
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:364]: why can we not call super()._process_view here?

a lot of this logic looks copy-pasted

File: metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py [hunk 3]
```diff
@@ -426,37 +506,49 @@ def loop_projections(
...
+                                aspectName="upstreamLineage",
+                                aspect=upstream_lineage,
+                            )
+
+                            lineage_wu = MetadataWorkUnit(
+                                id=f"{self.platform}-{lineage_mcpw.entityUrn}-{lineage_mcpw.aspectName}",
+                                mcp=lineage_mcpw,
+                            )
+
+                            self.report.report_workunit(lineage_wu)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:532]: why was this stuff changed - the as_workunit code was cleaner

File: metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py [hunk 4]
```diff
@@ -537,49 +629,83 @@ def _process_projections(
...
+                )
+                continue
+            self.report.report_entity_profiled(dataset_name)
             logger.debug(
-                f"Encountered ProgrammingError. Retrying with quoted schema name for schema {schema} and table {properties} %s",
-                error,
+                f"Preparing profiling request for {schema}, {projection}, {partition}"
+            )
+
+            yield GEProfilerRequest(
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:695]: what changed in here vs the superclass? please add it as a comment, and also consider if we can avoid the duplicate code

File: metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py [hunk 5]
```diff
@@ -614,13 +740,70 @@ def loop_models(
...
+                    )
+                    mce = MetadataChangeEvent(proposedSnapshot=dataset_snapshot)
+                    yield SqlWorkUnit(id=dataset_name, mce=mce)
+                    dpi_aspect = self.get_dataplatform_instance_aspect(
+                        dataset_urn=dataset_urn
                     )
+                    if dpi_aspect:
+                        yield dpi_aspect
+                    yield MetadataChangeProposalWrapper(
+                        entityType="dataset",
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:793]: only entityUrn and aspect args are required to `MetadataChangeProposalWrapper` - please remove the other ones

Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:793]: also let's keep this logic in _process_model?

File: metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py [hunk 6]
```diff
@@ -1123,115 +993,24 @@ def _get_upstream_lineage_info_projection(
...
-            get_owner_query = sql.text(
-                dedent(
-                    """
-                    SELECT owner_name
-                    FROM v_catalog.projections
-                    WHERE projection_name = '%(projection)s'
-                    """
-                    % {"projection": table}
-                )
+            columns = inspector.get_all_columns(table, schema)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:1008]: was there a reason to call this get_all_columns and not get_columns?

**Review by hsheth2 — COMMENTED** [review]
overall looks better than before

- still worried about the copy-pasted code here - I would like to minimize that if possible
- the profiling changes look suspect
- I haven't reviewed the tests yet

File: metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py (hunk 6 above)
Comment (vishalkSimplify) [reply metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:1008]: get_columns is used by profiling and its query are table level. We had to create a new function with schema level query .

File: metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py (hunk 1 above)
Comment (vishalkSimplify) [reply metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:315]: done

File: metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py (hunk 3 above)
Comment (vishalkSimplify) [reply metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:532]: reverted back to as_workunit

File: metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py (hunk 4 above)
Comment (vishalkSimplify) [reply metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:660]: added comments, main changes are we added one more loop in the function which gets projections and done profiling for projection. all logics and other things are same .

File: metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py (hunk 2 above)
Comment (vishalkSimplify) [reply metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:364]: we had to use copy-pasted code due to the columns variable, we cant use get_columns as it fetched queries based on columns. we had to create a new function for columns and properties. and since it is not possible to override the parent class variable data. we were not able to use super here and due to this all copy-pasted code.

**Comment by vishalkSimplify** [comment]
> Overall looks better than before
> 
> * still worried about the copy-pasted code here - I would like to minimize that if possible
> * the profiling changes look suspect
> * I haven't reviewed the tests yet

All the possible efforts are done for reducing the copy-pasted code. There were some reasons for every copy pasted code,i have replied with reasons in the comments below.

profiling - we have added comments in profiling functions .

File: metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py (hunk 6 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:990]: Critically: `get_columns` should still do the right thing, and do it efficiently.

Comment (vishalkSimplify) [reply metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py:990]: It does not . Get_all_columns is duplicate of get_columns . The reason for duplication was that profiling also calls get_columns . If we used get_columns call get_all_coulmns . Profiling will also call it  . We wanted to keep profiling totally seperate from other features of plugin .  Please check get_all_columns here - https://github.com/vertica/vertica-sqlalchemy-dialect/blob/main/vertica_sqlalchemy_dialect/base.py .

Comment (vishalkSimplify) [reply metadata-ingestion/src/datahub/ingestion/source/sql/vertica.py]: resolved , as we removed get_all_columns and used get_columns

**Review by hsheth2 — APPROVED** [review]
Lgtm. Thanks for all the work during reviews.

Will merge once CI is green
